File: leetcode/cpp/75.sort-colors.cpp
#include "oj_header.h"

class Solution {
public:
    void sortColors(vector<int>& nums)
    {
        // 两个指针分别指向 0 || 2 当前的边界
        int zero = -1; //左边及 zero 都是0
        int two = nums.size(); //右边及two都是2
        int one = 0;

        while (one < two) {
            if (nums[one] == 0) {
                // 左边是00011这种
                // 将一个1 与 这个0交换
                ++zero;
                swap(nums[zero], nums[one]);
                ++one;
            } else {
                if (nums[one] == 2) {
                    //交换过来的可能是0 / 1 不能直接指向下一个节点
                    --two;
                    swap(nums[two], nums[one]);

                } else {
                    // 左边肯定都是1
                    ++one;
                }
            }
        }
    }
};

File: leetcode/python/75.sort-colors.py
from typing import List

class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if len(nums) <= 1:
            return
        left = 0
        right = len(nums) - 1
        for i in range(len(nums)):
            # left == right 时 [0:left-1] 为 0  [right+1:] 为 2
            # 此时 [0 , 1 , 2]都可以 可以直接退出循环了
            # left <= i <= right
            # left <= i 是总是满足的
            while nums[i] != 1 and left <= right and i <= right:
                if nums[i] == 0:
                    # [left,i-1]只可能是 1
                    nums[i], nums[left] = nums[left], nums[i]
                    left += 1
                    break
                else:
                    # right 所在可能为 0 1 22
                    nums[i], nums[right] = nums[right], nums[i]
                    right -= 1
                if left >= right or i > right:
                    break
            if i > right or left > right:
                break
